s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java

```diff
@@ -179,6 +180,26 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
     }
   }
 
+  private static final class S2AStubCleanupNegotiationHandler extends ProtocolNegotiationHandler {
+    private final S2AStub s2aStub;
+
+    private S2AStubCleanupNegotiationHandler(
+        ChannelHandler next,
+        ChannelLogger logger,
+        S2AStub s2aStub) {
+      super(next, logger);
+      this.s2aStub = s2aStub;
+    }
+
+    @Override
+    protected void protocolNegotiationEventTriggered(ChannelHandlerContext ctx) {
+      s2aStub.close();
+      fireProtocolNegotiationEvent(ctx);
+      ctx.pipeline().remove(this);
```

Comment:
I see now that `fireProtocolNegotiationEvent()` does this for us (via replace()).

Reply:
Done.